lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -1116,6 +1117,253 @@ static bool canonicalizeOrOfConcatsWithCstOperands(OrOp op, size_t concatIdx1,
   return true;
 }
 
+/// Identifies integer range checks and simplifies them to minimise comparisons.
+///
+/// In an 'or' operation, tests of a variable belonging to a set of disjoint
+/// intervals are identified and merged whenever possible. 3 distinct patterns
+/// are considered to identify the intervals:
+///
+///   - icmp(ult, x, n),  constrains x < n
+///   - icmp(eq, x, n), constrains x = n
+///   - and(icmp(ugt, x, n), icmp(ult, x, m)) constrains n < x < m
+///
+/// For an individual value x, all the comparisons are collected and new
+/// comparisons are inserted to represent the intervals with fewer operations.
+///
+/// If a sequence of equality checks or(x == n, x == n + 1, ...) is detected,
+/// an interval is emitted only when at least 3 equality checks are folded into
+/// two unsigned comparisons.
+///
+/// All constants and operations operate in unsigned mode.
+///
+static bool tryMergeRanges(OrOp op, PatternRewriter &rewriter) {
+  // x == value
+  struct EqCheck {
+    APInt value;
+    Value check;
+  };
+  // start < x < end
+  struct RangeCheck {
+    APInt start;
+    APInt end;
+    Value startCheck;
+    Value endCheck;
+  };
+  // x < value
+  struct BoundCheck {
+    APInt value;
+    Value check;
+  };
+
+  using AnyCheck = std::variant<RangeCheck, EqCheck, BoundCheck>;
+
+  // Identify all the relevant patterns among the inputs,
+  // mark all others to be retained unchanged.
+  auto inputs = op.inputs();
+  unsigned numKeptOperands = 0;
+  std::vector<bool> keptOperands(inputs.size());
+  DenseMap<Value, SmallVector<std::pair<unsigned, AnyCheck>>> argChecks;
+
+  for (unsigned i = 0, n = inputs.size(); i < n; ++i) {
+    auto input = inputs[i];
+
+    // Find bound checks: and(x > n, x < m)
+    auto andOp = input.getDefiningOp<AndOp>();
+    if (andOp && andOp.inputs().size() == 2) {
+      ICmpOp lhsOp = andOp.inputs()[0].getDefiningOp<ICmpOp>();
+      ICmpOp rhsOp = andOp.inputs()[1].getDefiningOp<ICmpOp>();
+
+      if (lhsOp && rhsOp && lhsOp.lhs() == rhsOp.lhs()) {
+        Value arg = lhsOp.lhs();
+
+        APInt lhsBound, rhsBound;
+        if (matchPattern(lhsOp.rhs(), m_RConstant(lhsBound)) &&
+            matchPattern(rhsOp.rhs(), m_RConstant(rhsBound))) {
+          ICmpPredicate lhsPred = lhsOp.predicate();
+          ICmpPredicate rhsPred = rhsOp.predicate();
+          if (lhsPred == ICmpPredicate::ugt && rhsPred == ICmpPredicate::ult) {
+            argChecks[arg].emplace_back(i,
+                RangeCheck{lhsBound, rhsBound, lhsOp, rhsOp});
+            continue;
+          }
+          if (lhsPred == ICmpPredicate::ult && rhsPred == ICmpPredicate::ugt) {
```

Comment:
Do we want to consider signed comparisons here?

Reply:
At this point I'd say no. The point of this canonicalizer is to fold sequences of x == n, x == n + 1, ....
Since the ordering of canonicalizers is not determined, the one that folds a tree of ORs into a variadic OR can run intermixed with this one, creating partial intervals which later need to be merged.
Currently, the drawback is with intervals that cross 0, however they seem to be rare. In such cases, a separate canonicalizer could turn the unsigned ranges into signed ones.